lm_eval/models/huggingface.py

```diff
@@ -678,39 +678,61 @@ def _create_tokenizer(
+    def _detect_batch_size(self, requests=None, pos: int = 0) -> int:
+        if len(requests[0]) == 3:  # logprob evals
             _, context_enc, continuation_enc = requests[pos]
             max_length = len(
                 (context_enc + continuation_enc)[-(self.max_length + 1) :][:-1]
             )
             max_context_enc = len(context_enc[-(self.max_length + 1) :])
             max_cont_enc = len(continuation_enc[-(self.max_length + 1) :])
-        else:
-            max_length = self.max_length
+            security_margin_factor = (
+                4  # batch sizes for log prob evals sometimes generate OOMs
+            )
+        elif len(requests[0]) == 2:  # generative evals
+            # using rolling window with maximum context
+            longest_context = max(
+                [
+                    len(self.tok_encode(request[0]))
+                    + request[1].get("max_gen_toks", self.max_length)
+                    for request in requests[pos:]
+                ]
+            )
+            if longest_context > self.max_length:
+                eval_logger.warning(
+                    f"Longest context length of {longest_context} exceeds max_length of {self.max_length}. Truncating to max_length."
+                )
+                longest_context = self.max_length
+            max_length = longest_context
             max_context_enc = max_length
             max_cont_enc = max_length
+            security_margin_factor = 4
 
         # if OOM, then halves batch_size and tries again
         @find_executable_batch_size(starting_batch_size=self.max_batch_size)
         def forward_batch(batch_size):
+            security_margin = int(0.05 * security_margin_factor * batch_size)
             if self.AUTO_MODEL_CLASS == transformers.AutoModelForSeq2SeqLM:
                 length = max(max_context_enc, max_cont_enc)
                 batched_conts = torch.ones(
-                    (batch_size, length), device=self.device
+                    (batch_size + security_margin, length), device=self.device
+                ).long()
+                test_batch = torch.ones(
+                    (batch_size + security_margin, length), device=self.device
                 ).long()
-                test_batch = torch.ones((batch_size, length), device=self.device).long()
                 call_kwargs = {
                     "attn_mask": test_batch,
                     "labels": batched_conts,
                 }
             else:
                 call_kwargs = {}
-                test_batch = torch.ones(
-                    (batch_size, max_length), device=self.device
+                test_batch = torch.rand(
+                    (batch_size + security_margin, max_length), device=self.device
                 ).long()
-            for _ in range(5):
-                out = F.log_softmax(self._model_call(test_batch, **call_kwargs), dim=-1)  # noqa: F841
+
+            for _ in range(5 * security_margin_factor):
```

Comment:
Was 5 model calls, now it is 20. Why? It is time consuming, and may be money consuming.
May also be another constant (or may be even variable for all lm_eval to be controlled with args with backwards compatible defaults)

Reply:
We observed a lot of OOMs when using the auto batch size as was, especially for infinite context model types - some of which we observed being mitigated by increasing the number of loops here.
I'm OK with reverting this change, or adding the security margin as a parameter somewhere that we could increase for the leaderboard runs